eurocodedesign/units/__init__.py

```diff
@@ -0,0 +1,255 @@
+from abc import ABC
+from enum import Enum, unique, auto
+from functools import partial
+from typing import Self, TypeAlias
+
+
+@unique
+class PhysicalType(Enum):
+    DIMENSIONLESS = auto()
+    ANGLE = auto()
+    TIME = auto()
+    LENGTH = auto()
+    AREA = auto()
+    VOLUME = auto()
+    SECOND_MOMENT_OF_AREA = auto()
+    SPEED = auto()
+    ACCELERATION = auto()
+    MASS = auto()
+    FORCE = auto()
+    PRESSURE = auto()
+    ENERGY = auto()
+    TEMPERATURE = auto()
+
+
+class Prefix(Enum):
+    "Metric prefixes from nano (1e-9) to giga (1e9)"
+    giga = G = 1e9
+    mega = M = 1e6
+    kilo = k = 1e3
+    hecto = h = 1e2
+    deca = da = 1e1
+    none = one = 1e0
+    deci = d = 1e-1
+    centi = c = 1e-2
+    milli = m = 1e-3
+    micro = µ = 1e-6
+    nano = n = 1e-9
+
+
+_allowed_multiplications = {}
+
+
+class AbstractUnit(ABC):
+    _value: float = 1.0
+    _prefix: Prefix = Prefix.none
+    _unit: Self
+    _power = 1
+    _physical_type: PhysicalType
+
+    def _is_prefix_allowed(self, type: PhysicalType):
+        return type not in (PhysicalType.TIME,
+                            PhysicalType.ANGLE,
+                            PhysicalType.MASS,
+                            PhysicalType.TEMPERATURE)
+
+    def __init__(self, value: float = 1.0, prefix: Prefix = Prefix.none):
+        if (prefix != Prefix.none and
+            not self._is_prefix_allowed(prefix)):
+            raise ValueError(f'Metric prefix {prefix} not supported'
+                             f' for type {type(self)}')
+        self._unit = self.__class__
+        self._prefix = prefix
+        self._value = value * float(prefix.value ** self._power)
+
+    def __str__(self):
+        value = self._value / (self._prefix.value ** self._power)
+        prefix = self._prefix.name if self._prefix != Prefix.none else ''
+        return f"{value} {prefix}{self._unit.__name__}"
+
+    def __repr_latex_(self):
+        raise NotImplementedError
+
+    def __add__(self, other: Self):
+        if (type(self) != type(other) or
+            self._physical_type != other._physical_type):
```

Comment:
Doesn't `(type(self) != type(other)` cover `self._physical_type != other._physical_type` making the `or` unnecessary. 
Thia is based on the fact that `_physical_type` is property of the unit.

Reply:
It's necessary, because self and other can both be AbstractUnit, but have a different _physical_type.